Next 13

kristojorg opened the pull request:
This just bumps us up to Next 13 without using any of the new experimental features, since we are upgrading anyways

oliviawongnyc reviewed (commented):
Just a couple of small comments. Are the test failures expected?

kristojorg commented on src/components/chakra-ui/AuthButton.tsx:
```diff
@@ -38,7 +38,7 @@ export const AuthButton = forwardRef<HTMLButtonElement, AuthButtonProps>(
       </Button>
     ) : (
-      <Link href={SIGNIN_PICKER_PATH} passHref>
+      <Link href={SIGNIN_PICKER_PATH} passHref legacyBehavior>
```
I did these manually, I wanted to see if there was anywhere we could just remove the `<a>` tag beneath easily but there were only a couple places

kristojorg commented on src/components/chakra-ui/header/MenuLink.tsx:
```diff
@@ -3,16 +3,26 @@ import { useRouter } from "next/router";
-        <HStack spacing={3} as="a" py={4} px={8} sx={menuLinkStyle(isSelected)}>
+        <HStack
+          spacing={3}
+          as="span"
```
Good question - I will check if that's the default

oliviawongnyc replied:
Ah, thank you!

kristojorg commented:
I have a couple conflicts to resolve, but this is mostly ready for code review. Will resolve conflicts soon!

Toxiapo commented on src/components/chakra-ui/BookCard.tsx:
```diff
@@ -56,7 +56,6 @@ function BookCover({ book }: { book: SimpleBook }) {
       />
       <LazyImage
         ratio={0.7}
-        width="100vw"
```
We might need to bring this back. The images on the carousel show different sizes.

kristojorg replied:
Ohhh shoot this is in use in the carousel whoops

kristojorg replied:
Ok I cleaned this up. Ended up getting rid of LazyImage because it's not really necessary anymore, we can just use next/image directly. In order to make things work better, I switched the carousel books to be sized with flex instead of width, which seems to be working.

In all, switching to next/image seems to cut the image weights 50% over the (now older) next/legacy/image. Also should cut down on JS size and increase perf because it uses browser apis for lazy loading

Toxiapo replied:
Nice!!

Toxiapo reviewed (commented):
Looks great! Just some small questions.

I am also curious if it's painful to upgrade to nextjs? It's a very big change.

kristojorg commented on src/components/chakra-ui/BookDetail/index.tsx:
```diff
@@ -114,11 +114,12 @@ export const BookDetails: React.FC<BookDetailsProps> = ({
-                objectFit="contain"
+                height={360}
+                width={280}
+                style={{ objectFit: "contain", height: "360px" }}
```
I thought so too, but it appears that `height` and `width` on an image are used to figure out the aspect ratio and the sizes, but not actually used to style the image. I'm not sure if this is a particularity (or bug) in the next/image component or if it's how the `height`/`width` props are meant to work on images. 

tldr: I had to set both in order to make the image render at that height

kristojorg commented on tsconfig.json:
```diff
@@ -17,7 +17,12 @@
     "baseUrl": "src",
-    "incremental": true
+    "incremental": true,
+    "plugins": [
+      {
+        "name": "next"
+      }
```
I actually don't know how that got there. Will remove

kristojorg replied:
Actually! I removed it and it came back 🤔 . Then I randomly saw this twee: https://twitter.com/delba_oliveira/status/1592209522281439232?s=20&t=3xsdt_wA5cLLVahlEuYENw

Looks like it's a plugin that brings NextJS documentation into VSCode !

Toxiapo replied:
ohhh this is useful!

kristojorg commented on src/components/chakra-ui/BookCard.tsx:
```diff
@@ -45,7 +59,13 @@ const BookCard = React.forwardRef<
+      position="relative"
+      height={0}
+      width="100%"
+      paddingTop="143%"
```
Good question. Previously we used an aspect ratio of 0.7 and this is the inverse (1/0.7). I should add a comment making that clear.

kristojorg commented:
> I am also curious if it's painful to upgrade to nextjs? It's a very big change.

This one wasn't too painful honestly, only a couple APIs changed so it was pretty seamless

samanthaandrews commented:
I'd like to get this merged soon. Looks like we have some build failures. What can I do to help?

kristojorg commented:
@samanthaandrews Ok Next published a patch release that was actually a breaking change in how they worked with cookies in middleware 😠 . I have updated our middleware to work properly now, and also updated the middleware tests. I basically removed the `getCredentials` mock and instead set cookies on the mock `NextRequest` we instantiated. 

Should be ready to go now.

Just need to rebase my monorepo on this next...

samanthaandrews reviewed (approved):
LGTM!

Toxiapo reviewed (approved):
Looks great!
